### DISCUSSION TITLE
Can i manually limit the value of a variable to a specific range?

### DISCUSSION THREAD

**Discussion by leezw37** [idaholab/moose#22792] [Q&A General] [answered]
Hi guys, I'm trying to create my own application for two-phase flow problems based on MOOSE. And I need to limit the value of a solution variable to a specific range. (volume fraction of vapor is between 0 and 1) But i don't know how. 

Can anybody help me? Thanks!

**Comment by WilkAndy** [comment]
Not to answer your question, but i'm just curious why the PorousFlow module of MOOSE, which has multi-phase flow (eg gas + water): https://mooseframework.inl.gov/modules/porous_flow/index.html , doesn't suit you.  I like to try to keep in touch with what people need.

**Reply by leezw37** [reply]
Thanks for your reply. I'm not familiar with PorousFlow module, and i'll have a look.

I want to build a simplified application of RELAP-7: https://moose.inl.gov/relap7/SitePages/Home.aspx. 
And thermal_hydraulics module is similar to the single phase part of RELAP-7: https://mooseframework.inl.gov/modules/thermal_hydraulics/index.html

**Comment by GiudGiud** [comment, marked as answer]
Hello

For limiting variable values, you may use the Bounds system
https://mooseframework.inl.gov/source/auxkernels/ConstantBoundsAux.html
Good luck with this effort!

Guillaume

**Reply by leezw37** [reply]
Oh man, you're amazing! It works fine according to my test! 

(btw, i misunderstood the meaning of HeatConductionMaterial.C:computeProperties(). The temp limit is just for calculating other material properties, so i edit the question to make my question clear. )

Thank you so much!

**Reply by WilkAndy** [reply]
I'm so pleased it works for you, @leezw37 .   Bounds are quite difficult in practice.

**Reply by hugary1995** [reply]
While this may work for you, below are several things to consider before you use it in production (also to remind future readers of this thread).

The most typical use of the bound system is probably phase field problems, where the order parameter/phase/volume fraction is bounded. These systems are solving for necessary conditions of a minimization problem in nature, so if you find yourself needing to set bound on the primary variable, you are likely facing one of the following issues:
1. The objective function (energy) is concave in the primary variable.
2. The objective function is globally convex and has a unique minimum close to the bound.
3. The objective function is non-convex, but it is locally convex in the neighborhood of the solution.
4. The objective function has several local minima, and by applying the bound, nonphysical minima are ruled out.

You really need to think about which case your problem falls into, as bounding the variable isn't the solution to all.
1. You can see if it is case 1 by looking at the singular values, If indeed it is, then there is no point in enforcing any bounds.
2. This is a perfectly valid case for the use of bounds, as numerically (e.g. in Newton-Raphson iterations) the solution may go outside the physical bounds, and some of your constitutive definition may become invalid.
3. Enforcing a bound may help, but you are also at the risk of finding the wrong solution. For a transient problem, you can often get around this by cutting the time step, assuming that the solution from the previous step is correct.
4. This is also a valid use case for bounds, but the rate of convergence may be compromised.

**Reply by leezw37** [reply]
Thank you, Hu.  Your reply is much more detailed, and very helpful to me!

As what you said, I actually don't need bounds to my simulation temporarily. And sometimes the bounds make the solution difficult to solve. 
For the trouble with my application, the solution is to make some reasonable changes to the closure relations and they work well without additional bounds now. But i may use it when i need it.

Anyway, thank you very much. You are impressive. Have a nice day!
